### ISSUE TITLE
Recipes for previous versions should be retained

### ISSUE THREAD

**Issue by moorepants** [conda/conda-recipes#456]
It seems like recipes get updated here when new versions come out, but there are a number of use cases for building previous package versions. For example, in a number of software packages that I work with we run continuous integration on sets of the minimum deps, latest versions, and the development versions of packages. Or I need to build very specific package combinations for research reproducibility's sake. If the old versions of recipes are hidden in the git history it makes it difficult for the recipes to be found, especially since there are no tags that pin commits to recipe versions.

My opinion is that each directory in this repo should be more like:

```
conda-recipes
-> numpy/
    --> 1.9.2/
    --> 1.10.1/
    --> 1.10.1-openblas/
    --> 1.10.1-mkl/
    --> etc
```

So each package will have all of the working recipes for any version, any features, or any special combination of dependencies.

Unless the recipes can be generalized across versions, it seems like retaining all of the previous versions is a good idea.

**Comment by patricksnape** [comment]
All the version revisions are in the git history. So I don't think copying all of the recipes is a good idea. However, maybe tagging the releases would be good so that you can easily refer to the recipe version without having to resort to an ugly SHA.

**Comment by msarahan** [comment]
Tagging is not really an option on conda-recipes IMHO since the tags really
should be per recipe, not at the repository level.

**Comment by patricksnape** [comment]
@msarahan I totally agree - but that isn't possible with the current monolithic recipe layout :) Another reason why it would be great to break this repo apart IMHO. However, spilling all recipes out into folders could get crazy for a project that updates really quickly. Plus, what if there is an error in the recipe? Then you would have to change it in many places at once and one may get forgotten. I personally just prefer simplicity and potentially hundreds of versions across hundreds of projects in one repository starts to feel really overwhelming.

**Comment by moorepants** [comment]
Tagging is not out of the question here. If the various versions had git tags, you simply have to understand that the git tag only corresponds to the referenced directory and is meaningless to the rest of the repo.

In general, I always found this repo odd. It seems to me that the core conda recipes should belong with the software's repository, just like a setup.py file, CMAKE files, or a debian directory. That way the recipe always matches the source code. The next level could be more specific build recipes that tie to a specific anaconda channel. These may vary from the conical recipes and can be used to reproduce the contents of a channel. But this giant repo of random recipes isn't always that useful.

**Comment by aterrel** [comment]
It's a matter of ease. 

Can you imagine trying to get all these recipes in each source repo? Then should those repos have the spec files for rpm, deb, pacman, homebrew, easybuild, hashdist, buildsystem, chocolatey, and so on?

This is pretty typical for package management systems to have a large repo of package specifications. In other repos like this that I've worked with I have seen the spec version in the name, then you don't have to have a thousand tags to sort through or git bisect to find when the version changed.
